### PR TITLE
Feature Facebook sign in

### PR DESCRIPTION
**Pull request by jpascual1994**
This PR includes:
- Sign in with facebook and tests

Trello card: https://trello.com/c/nLmdJdix/1-2-as-a-user-i-should-be-able-to-login-with-facebook-so-i-can-register-faster

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,6 +1,22 @@
...
+
+  FACEBOOK_GENDERS = { 'male' => 1, 'female' => 2 }
+
+  def self.from_omniauth(auth)
+    where( provider: auth.provider, uid: auth.uid ).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
+      user.name = auth.info.name
+      if FACEBOOK_GENDERS.key?( auth.extra.raw_info.gender )
+        user.gender = FACEBOOK_GENDERS[auth.extra.raw_info.gender]
```
Comment (gabrielrp) [inline app/models/user.rb]: Declare `gender` as `enum` and then check if the facebook gender matches any of ours.

Comment (gabrielrp) [reply app/models/user.rb]: Also, write it in one line, something like `user.gender = genders.key(auth.extra.raw_info.gender) ? auth.extra.raw_info.gender : 'others'`

**Review by gabrielrp — CHANGES_REQUESTED** [review]

File: config/initializers/devise.rb [hunk 2]
```diff
@@ -250,6 +250,8 @@
   # ==> OmniAuth
   # Add a new OmniAuth provider. Check the wiki for more information on setting
   # up on your models and hooks.
+  config.omniauth :facebook, '1957979147801504', '6e06570424d211750ca738deb028a9ee',
```
Comment (mdornel) [inline config/initializers/devise.rb]: You shouldn't publish your keys in github, this is private data. Use `figaro` gem instead (check https://github.com/laserlemon/figaro)

You should create a file using the gem, for example:
```
# config/application.yml

FACEBOOK_APP_ID: "YOUR_APP_ID"
FACEBOOK_APP_KEY: "YOUR_FACEBOOK_APP_KEY"
```

And use it as: `ENV['FACEBOOK_APP_ID']`

File: spec/features/facebook_log_in_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    Rails.application.env_config['omniauth.auth'] = OmniAuth.config.mock_auth[:facebook]
+  end
+
+  context 'valid login with facebook' do
+    background do
+      set_valid_omniauth()
+      visit '/'
+      click_link 'Connect with facebook'
+      sleep 1
+      visit '/users/auth/facebook'
```
Comment (mdornel) [inline spec/features/facebook_log_in_spec.rb]: Here you can use the path helper for this route.

File: spec/features/facebook_log_in_spec.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    scenario 'show the user name' do
+      expect(page).to have_content( User.last.name)
+    end
+  end
+
+  context 'invalid login with facebook' do
+    scenario 'redirect to root' do
+      set_invalid_omniauth()
+      visit '/'
```
Comment (mdornel) [inline spec/features/facebook_log_in_spec.rb]: Here too.

**Review by mdornel — CHANGES_REQUESTED** [review]

**Review by mdornel — APPROVED** [review]

File: spec/support/omniauth_macros.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    OmniAuth.config.test_mode = true
+    OmniAuth.config.mock_auth[:facebook] = :invalid_credentials
+  end
+
+  def set_request_env_for_omniauth
+    request.env["devise.mapping"] = Devise.mappings[:user]
+    request.env["omniauth.auth"] = OmniAuth::AuthHash.new( valid_response )
+  end
+
+  def valid_response()
```
Comment (gabrielrp) [inline spec/support/omniauth_macros.rb]: The parentheses aren't necessary

File: spec/support/omniauth_macros.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
+module OmniauthMacros
+  def set_valid_omniauth(opts = {})
```
Comment (gabrielrp) [inline spec/support/omniauth_macros.rb]: Where do you use the `opts` param?

File: spec/support/omniauth_macros.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    OmniAuth.config.test_mode = true
+    OmniAuth.config.add_mock( :facebook, valid_response )
+  end
+
+  def set_invalid_omniauth(opts = {})
```
Comment (gabrielrp) [inline spec/support/omniauth_macros.rb]: Same^

File: spec/features/facebook_log_in_spec.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+feature 'Confirm Account', js: true do
+  background do
+    Rails.application.env_config['devise.mapping'] = Devise.mappings[:user]
+    Rails.application.env_config['omniauth.auth'] = OmniAuth.config.mock_auth[:facebook]
+  end
+
+  context 'valid login with facebook' do
+    background do
+      set_valid_omniauth()
```
Comment (gabrielrp) [inline spec/features/facebook_log_in_spec.rb]: Parentheses aren`t necessary

File: spec/features/facebook_log_in_spec.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    end
+
+    scenario 'show the user name' do
+      expect(page).to have_content( User.last.name)
+    end
+  end
+
+  context 'invalid login with facebook' do
+    scenario 'redirect to root' do
+      set_invalid_omniauth()
```
Comment (gabrielrp) [inline spec/features/facebook_log_in_spec.rb]: Same^

File: spec/features/facebook_log_in_spec.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      sleep 1
+      visit user_facebook_omniauth_authorize_path
+    end
+
+    scenario 'redirect to homepage' do
+      expect(page).to have_current_path( homepage_users_path )
+    end
+
+    scenario 'show the user name' do
+      expect(page).to have_content( User.last.name)
```
Comment (gabrielrp) [inline spec/features/facebook_log_in_spec.rb]: Remove extra space

**Review by gabrielrp — CHANGES_REQUESTED** [review]

**Review by gabrielrp — APPROVED** [review]

**Pull request merged**